scipy/stats/_binomtest.py

```diff
@@ -0,0 +1,258 @@
+                                              self.alternative)
+        else:
+            # method is 'wilson' or 'wilsoncc'
+            low, high = _binom_wilson_conf_int(self.k, self.n,
+                                               confidence_level,
+                                               self.alternative,
+                                               correction=method == 'wilsoncc')
+        return ConfidenceInterval(low=low, high=high)
+
+
+def _binom_exact_conf_int(k, n, confidence_level, alternative):
+    """
+    Compute the estimate and confidence interval for the binomial test.
+
+    Returns proportion, prop_low, prop_high
+    """
+    if alternative == 'two-sided':
+        alpha = (1 - confidence_level) / 2
+        if k == 0:
+            plow = 0.0
+        else:
+            plow = brentq(lambda p: binom.sf(k-1, n, p) - alpha, 0, 1)
+        if k == n:
+            phigh = 1.0
+        else:
+            phigh = brentq(lambda p: binom.cdf(k, n, p) - alpha, 0, 1)
+    elif alternative == 'less':
+        alpha = 1 - confidence_level
+        plow = 0.0
+        if k == n:
+            phigh = 1.0
+        else:
+            phigh = brentq(lambda p: binom.cdf(k, n, p) - alpha, 0, 1)
+    elif alternative == 'greater':
+        alpha = 1 - confidence_level
+        if k == 0:
+            plow = 0.0
+        else:
+            plow = brentq(lambda p: binom.sf(k-1, n, p) - alpha, 0, 1)
+        phigh = 1.0
+    return plow, phigh
+
+
+def _binom_wilson_conf_int(k, n, confidence_level, alternative, correction):
+    # This function assumes that the arguments have already been validated.
+    # In particular, `alternative` must be one of 'two-sided', 'less' or
+    # 'greater'.
+    p = k / n
+    if alternative == 'two-sided':
+        z = ndtri(0.5 + 0.5*confidence_level)
+    else:
+        z = ndtri(confidence_level)
+
+    t = 1 + z**2/n
+    r = (p + z**2/(2*n)) / t
+
+    if correction:
+        if alternative == 'less' or k == 0:
+            lo = 0.0
+        else:
+            dlo = ((z * sqrt(z**2 - 1/n + 4*n*p*(1 - p) + (4*p - 2)) + 1) /
+                   (2*n*t))
+            lo = r - dlo
+        if alternative == 'greater' or k == n:
+            hi = 1.0
+        else:
+            dhi = ((z * sqrt(z**2 - 1/n + 4*n*p*(1 - p) - (4*p - 2)) + 1) /
+                   (2*n*t))
+            hi = r + dhi
+    else:
+        d = z/t * sqrt(p*(1-p)/n + (z/(2*n))**2)
+        if alternative == 'less' or k == 0:
+            lo = 0.0
+        else:
+            lo = r - d
+        if alternative == 'greater' or k == n:
+            hi = 1.0
+        else:
+            hi = r + d
+
+    return lo, hi
+
+
+def binomtest(k, *, n=None, p=0.5, alternative='two-sided'):
+    """
+    Perform a test that the probability of success is p.
+
+    This is a test of the null hypothesis that the probability of success
+    in a Bernoulli experiment is `p`.
+
+    Parameters
+    ----------
+    k : int or array_like
```

Comment:
```
binomtest(k=0.2, n=2, p=0.9)
Out[7]: BinomTestResult(k=0, n=2, alternative='two-sided', proportion_estimate=0.0, pvalue=0.009999999999999995)
```
Looks like it rounds floats. Is that desired?

Reply:
```
res = binomtest(k=3, n=30, p=0.4)
res.proportion_ci(-1)
Out[9]: ConfidenceInterval(low=1.0, high=0.0)
res.proportion_ci(-2)
Traceback (most recent call last):

  File "<ipython-input-10-88be8049960e>", line 1, in <module>
    res.proportion_ci(-2)

  File "/Users/matthaberland/Desktop/scipy/scipy/stats/_binomtest.py", line 63, in proportion_ci
    self.alternative)

  File "/Users/matthaberland/Desktop/scipy/scipy/stats/_binomtest.py", line 84, in _binom_exact_conf_int
    plow = brentq(lambda p: binom.sf(k-1, n, p) - alpha, 0, 1)

  File "/Users/matthaberland/Desktop/scipy/scipy/optimize/zeros.py", line 776, in brentq
    r = _zeros._brentq(f, a, b, xtol, rtol, maxiter, args, full_output, disp)

ValueError: f(a) and f(b) must have different signs
```

Reply:
In the last update, I added more validation of the arguments:

```
In [2]: binomtest(2.2, n=10, p=0.3)                                                                                      
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-54590ac8b247> in <module>
----> 1 binomtest(2.2, n=10, p=0.3)

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+a843309-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_binomtest.py in binomtest(k, n, p, alternative)
    276 
    277     """
--> 278     k = _validate_nonneg_int(k, 'k')
    279     n = _validate_nonneg_int(n, 'n')
    280     if k > n:

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+a843309-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_binomtest.py in _validate_nonneg_int(k, name)
    191         k = operator.index(k)
    192     except TypeError:
--> 193         raise TypeError(f'{name} must be an integer.') from None
    194     if k < 0:
    195         raise ValueError(f'{name} must be nonnegative') from None

TypeError: k must be an integer.
```

```
In [3]: binomtest(-2, n=10, p=0.3)                                                                                       
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-dbf4da05bbf2> in <module>
----> 1 binomtest(-2, n=10, p=0.3)

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+a843309-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_binomtest.py in binomtest(k, n, p, alternative)
    276 
    277     """
--> 278     k = _validate_nonneg_int(k, 'k')
    279     n = _validate_nonneg_int(n, 'n')
    280     if k > n:

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+a843309-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_binomtest.py in _validate_nonneg_int(k, name)
    193         raise TypeError(f'{name} must be an integer.') from None
    194     if k < 0:
--> 195         raise ValueError(f'{name} must be nonnegative') from None
    196     return k
    197 

ValueError: k must be nonnegative
```

```
In [4]: binomtest(2, n=10, p=np.nan)                                                                                     
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-f4f8b300da16> in <module>
----> 1 binomtest(2, n=10, p=np.nan)

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+a843309-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_binomtest.py in binomtest(k, n, p, alternative)
    282 
    283     if not (0 <= p <= 1):
--> 284         raise ValueError("p must be in range [0,1]")
    285 
    286     if alternative not in ('two-sided', 'less', 'greater'):

ValueError: p must be in range [0,1]
```

```
In [5]: result = binomtest(3, n=15, p=0.25)                                                                              

In [6]: result.proportion_ci(confidence_level=-1)                                                                        
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-9c28ac142d7e> in <module>
----> 1 result.proportion_ci(confidence_level=-1)

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+a843309-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_binomtest.py in proportion_ci(self, confidence_level, method)
     99                              "'wilsoncc'.")
    100         if not (0 <= confidence_level <= 1):
--> 101             raise ValueError('confidence_level must be in the interval '
    102                              '[0, 1].')
    103         if method == 'exact':

ValueError: confidence_level must be in the interval [0, 1].
```

I didn't fix the edge case `k=0, n=0` yet.

Reply:
I think my latest update addresses this comment.

Reply:
It sure does!